File: aws/library/lib/src/main/java/pl/andrzejressel/deeplambdaserialization/aws/AbstractAWSLib.java
package pl.andrzejressel.deeplambdaserialization.aws;

import com.pulumi.asset.FileArchive;
import com.pulumi.core.Output;
import java.io.IOException;
import java.nio.file.Files;
import java.util.Base64;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;
import pl.andrzejressel.deeplambdaserialization.lib.DeepLambdaSerialization;
import pl.andrzejressel.deeplambdaserialization.lib.SerializableInputFunctionN;
import pl.andrzejressel.dto.serializator.Serializator;

class AbstractAWSLib {

  private static final String HANDLER =
      "pl.andrzejressel.deeplambdaserialization.aws.handler.Handler";

  protected static <RET> LambdaArguments handleFunctionN(
      SerializableInputFunctionN<RET> f, List<Output<?>> args) {

    @SuppressWarnings("unchecked")
    var castedArgs = args.stream().map(t -> (Output<Object>) t).collect(Collectors.toList());

    var envVars = Output.all(castedArgs).applyValue((list) -> {
      Map<String, String> envVarsMap = new HashMap<>();

      for (int i = 0; i < list.size(); i++) {
        var argument = (Object) list.get(i);
        @SuppressWarnings("unchecked")
        var serializator = (Serializator<Object>) f.getInputSerializators().get(i);

        envVarsMap.put(
            "ARG_" + (i + 1),
            Base64.getEncoder().encodeToString(serializator.serialize(argument).array()));
      }

      return envVarsMap;
    });

    var file = DeepLambdaSerialization.getJar(f);
    try {
      var dir = Files.createTempDirectory("deeplambdaserialization");
      Files.copy(file.openStream(), dir.resolve("test.jar"));
      var jarLocation = dir.resolve("test.jar");
      return new LambdaArguments(
          new FileArchive(jarLocation.toAbsolutePath().toString()), envVars, HANDLER);
    } catch (IOException e) {
      throw new RuntimeException(e);
    }
  }
}

File: aws/library/lib/src/main/java/pl/andrzejressel/deeplambdaserialization/aws/AbstractAWSLib.kt
// SPDX-License-Identifier: LGPL-3.0-or-later
package pl.andrzejressel.deeplambdaserialization.aws

import com.pulumi.asset.FileArchive
import com.pulumi.core.Output
import java.nio.file.Files
import java.util.Base64
import kotlin.io.path.absolutePathString
import pl.andrzejressel.deeplambdaserialization.lib.*
import pl.andrzejressel.dto.serializator.Serializator

@Suppress("UNCHECKED_CAST")
abstract class AbstractAWSLibKT {

  companion object {
    private val HANDLER = "pl.andrzejressel.deeplambdaserialization.aws.handler.Handler"

    @JvmStatic
    protected fun <RET> handleFunctionN(
        f: SerializableInputFunctionN<RET>,
        args: List<Output<*>>
    ): LambdaArguments {

      val anyArgs: List<Output<Any>> = args as List<Output<Any>>

      val envVars =
          Output.all(anyArgs).applyValue { list ->
            list
                .zip(f.inputSerializators)
                .mapIndexed { index, (obj, serializator) ->
                  serializator as Serializator<Any>

                  "ARG_${index + 1}" to
                      Base64.getEncoder().encodeToString(serializator.serialize(obj).array())
                }
                .toMap()
          }

      val file = DeepLambdaSerialization.getJar(f)

      val dir = Files.createTempDirectory("deeplambdaserialization")

      val fsJar =
          file.openStream().use {
            Files.copy(it, dir.resolve("test.jar"))
            dir.resolve("test.jar")
          }

      return LambdaArguments(FileArchive(fsJar.absolutePathString()), envVars, HANDLER)
    }
  }

  data class LambdaArguments(
      val code: FileArchive,
      val envVars: Output<Map<String, String>>,
      val handlerClass: String
  )
}
